### DISCUSSION TITLE
Proposal: extend NodeTypes structure to send props to the custom node

### DISCUSSION THREAD

**Discussion by raarts** [xyflow/xyflow#2084] [Ideas]
Hi,

I think that a lot of people want their custom nodes to update the data they are receiving. Current practice seems to be pass an `onUpdate` function in the data object. The problem is this is not actually part of the data the node is managing.

I think the proper place to put such an update function though is on the `NodeType` structure. And maybe other props for the node too. The current type is:

```
type NodeType = {
  {[p: string]: ReactNode}
}
```

so in my opinion something similar to the following would be an improvement:

```
type NodeInfo = {
  component: ReactNode;
  onUpdate: (id: string, data: any) => void;
  props: any;
}

type NodeType = {
  {[p: string]: NodeInfo}
}
```

or something similar. Naming could be improved. Please tell me what you think.

**Comment by alimoeeny** [comment]
or it if is easier to do, change the existing type def at
https://github.com/wbkd/react-flow/blob/3548891b8e4f6ffbb9de1fbbef582d8eb55304c7/src/types/general.ts#L12
to something like: 
```
export type NodeTypes = { [key: string]: ReactNode | (_:any)=> ReactNode };
```
